datafusion/physical-expr/src/window/lead_lag.rs

```diff
@@ -238,7 +238,10 @@ impl PartitionEvaluator for WindowShiftEvaluator {
             let start = if self.non_null_offsets.len() == self.shift_offset as usize {
                 // How many rows needed previous than the current row to get necessary lag result
                 let offset: usize = self.non_null_offsets.iter().sum();
-                idx.saturating_sub(offset + 1)
+                idx.saturating_sub(offset)
+            } else if !self.ignore_nulls {
+                let offset = self.shift_offset as usize;
+                idx.saturating_sub(offset)
```

Comment:
instead of 0, we now return `idx-offset` (can be larger than 0, which will trigger pruning)